src/System.Text.Json/System/Text/Json/JsonReader.cs

```diff
@@ -1,383 +1,836 @@
+        private readonly JsonEncoderState _encoderState;
+        private readonly TextEncoder _encoder;
+
+        private ReadOnlySpan<byte> _buffer;
+
+        // Depth tracks the recursive depth of the nested objects / arrays within the JSON data.
+        private int _depth;
+
+        // This mask represents a tiny stack to track the state during nested transitions.
+        // The first bit represents the state of the current level (1 == object, 0 == array).
+        // Each subsequent bit is the parent / containing type (object or array). Since this
+        // reader does a linear scan, we only need to keep a single path as we go through the data.
+        private ulong _containerMask;
+
+        // These next 2 properties are used to check for whether we can take the fast path
+        // for invariant UTF-8 or UTF-16 processing. Otherwise, we need to go through the
+        // slow path which makes use of the (possibly generic) encoder.
+        private bool UseFastUtf8 => _encoderState == JsonEncoderState.UseFastUtf8;
+        private bool UseFastUtf16 => _encoderState == JsonEncoderState.UseFastUtf16;
+
+        // These properties are helpers for determining the current state of the reader
+        private bool IsRoot => _depth == 1;
+        private bool InArray => (_containerMask & 1) == 0 && (_depth > 0);
+        private bool InObject => (_containerMask & 1) == 1;
+
+        /// <summary>
+        /// Gets the token type of the last processed token in the JSON stream.
+        /// </summary>
+        public JsonTokenType TokenType { get; private set; }
+
+        /// <summary>
+        /// Gets the value as a ReadOnlySpan<byte> of the last processed token. The contents of this
+        /// is only relevant when <see cref="TokenType" /> is <see cref="JsonTokenType.Value" /> or
+        /// <see cref="JsonTokenType.PropertyName" />. Otherwise, this value should be set to
+        /// <see cref="ReadOnlySpan{T}.Empty"/>.
+        /// </summary>
+        public ReadOnlySpan<byte> Value { get; private set; }
+
+        /// <summary>
+        /// Gets the JSON value type of the last processed token. The contents of this
+        /// is only relevant when <see cref="TokenType" /> is <see cref="JsonTokenType.Value" /> or
+        /// <see cref="JsonTokenType.PropertyName" />.
+        /// </summary>
+        public JsonValueType ValueType { get; private set; }
+
+        /// <summary>
+        /// Gets the encoder instance used when the reader was constructed.
+        /// </summary>
+        public TextEncoder Encoder => _encoder;
+
+        /// <summary>
+        /// Constructs a new JsonReader instance. This is a stack-only type.
+        /// </summary>
+        /// <param name="data">The <see cref="Span{byte}"/> value to consume. </param>
+        /// <param name="encoder">An encoder used for decoding bytes from <paramref name="data"/> into characters.</param>
+        public JsonReader(ReadOnlySpan<byte> data, TextEncoder encoder)
```

Comment:
Since the PR is ready and done, we would be throwing away work anyway, and so I think we should just merge. Possibly not that much will be thrown away, and we will be able to observe how support for multi-span buffers impacts perf.